### PR TITLE
[search] Tracer.

### PR DESCRIPTION
**Pull request by ygorshenin**
Tracer for search, collects all valid parses of the search query.

### CODE REVIEW

**Comment by mpimenov** [comment]
(discussed offline)
1. Only successful parses are emitted now. This is probably right, and emitting
the incomplete parses that is currently in the works will cover other cases.
EmitResult should still be the only point in the geocoder that emits results (and parses).
The search by lat-lon does not fit in here, but it is not important as the parses are trivial.
2. We probably should add a LogTracer that is easy to turn on and that would
print the parses to the console for debug purposes.

File: search/tracer.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+{
+class Tracer
+{
+public:
+  struct Parse
+  {
+    using TokenType = BaseContext::TokenType;
+
+    explicit Parse(std::vector<TokenType> const & types);
+    explicit Parse(std::vector<std::pair<TokenType, TokenRange>> const & ranges);
```
Comment (tatiana-kondakova) [inline search/tracer.hpp]: do we really need explicit here?

Comment (tatiana-kondakova) [reply search/tracer.hpp]: oops, sorry

File: search/tracer.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+#include <cstddef>
+#include <sstream>
+
+using namespace std;
+
+namespace search
+{
+// Tracer::Parse -----------------------------------------------------------------------------------
+Tracer::Parse::Parse(std::vector<TokenType> const & types)
```
Comment (tatiana-kondakova) [inline search/tracer.cpp]: `s/std::vector/vector`

Comment (ygorshenin) [reply search/tracer.cpp]: Done

**Comment by ygorshenin** [comment]
After discussion with @burivuh we concluded that it would be useful to emit matched categories for the innermost layer. IOW, for search request "cafe moscow" there can be two parses - ["poi", "city"] and ["poi"] (for the cafe named "Moscow"). In the former case it would be nice to somehow specify that this was not just "poi", but also a category "amenity-cafe".

**Comment by ygorshenin** [comment]
After discussion with @burivuh we concluded that it would be useful to emit matched categories for the innermost layer. IOW, for search request "cafe moscow" there can be two parses - ["poi", "city"] and ["poi"] (for the cafe named "Moscow"). In the former case it would be nice to somehow specify that this was not just "poi", but also a category "amenity-cafe".

**Comment by ygorshenin** [comment]
@burivuh - addition of the list of categories is not a big deal, and if you agree with the generic structure of the tracer - feel free to merge this PR, as @mpimenov works on geocoder concurrently.

**Pull request merged**